Add more tests for the dataframe interchange protocol

Rubtsowa opened the pull request:
Here we add generic tests for the df protocol. These tests can be used for any product that works with dataframes. It is only necessary to change conftest, replace the existing constructor for pandas with the one necessary for the project that will be used.

dchigarev commented on protocol/tests/test_protocol.py:
```diff
@@ -0,0 +1,142 @@
+        assert df2.get_column_by_name(column).null_count == 0
+        assert df2.get_column_by_name(column).size == column_size
+        assert df2.get_column_by_name(column).offset == 0
+        assert not df2["x"].is_masked
```
Does the [protocol](https://data-apis.org/dataframe-protocol/latest/API.html#interface) define `__getitem__` for `DataFrame` and `is_masked` for `Column`? I couldn't find any of these

rgommers replied:
It doesn't - `__getitem__` is avoided on purpose, because it is a heavily overloaded method with unclear semantics in existing dataframe libraries. Here we followed a "explicit is better than implicit" method, which makes sense for a protocol aimed at library developers rather than end users.

rgommers replied:
No `is_masked` either indeed.

dchigarev commented on protocol/tests/test_protocol.py:
```diff
@@ -0,0 +1,142 @@
+    assert df2.num_rows() == 3
+    assert df2.num_chunks() == 1
+    assert df2.column_names() == ["x", "y", "z"]
+    assert df2.select_columns((0, 2))._df[:, 0].tolist() == df2.select_columns_by_name(("x", "z"))._df[:, 0].tolist()
+    assert df2.select_columns((0, 2))._df[:, 1].tolist() == df2.select_columns_by_name(("x", "z"))._df[:, 1].tolist()
```
We can't use `._df` attribute of the `Column` object as it's an implementation-specific field. Could we do the same testing but with extracting the data with the protocol operations only (building buffers & comparing them)?

dchigarev commented:
@Rubtsowa could you please elaborate on what these tests are supposed to be testing? Going from what I see in the PR, this tests the protocol implementation of some pandas-like dataframe (the suites use pandas specific methods of `df` like [`df.columns.values`](https://github.com/Rubtsowa/dataframe-api/blob/0c70d7c3faecc3844a53e6ccfa25ca9d8da57bfc/protocol/tests/test_protocol.py#L21), [`df.categorize`](https://github.com/Rubtsowa/dataframe-api/blob/0c70d7c3faecc3844a53e6ccfa25ca9d8da57bfc/protocol/tests/test_protocol.py#L102), an assumption that [string column has 'np.object' type](https://github.com/Rubtsowa/dataframe-api/blob/0c70d7c3faecc3844a53e6ccfa25ca9d8da57bfc/protocol/tests/test_protocol.py#L96), etc). 

Is there a plan to make these tests more generic so not only pandas-like dataframes could be tested with them?

rgommers commented:
Thanks for the tests @Rubtsowa and for the initial review @dchigarev! @Rubtsowa it would be nice to fill in the PR description a bit, e.g., what are the main things this PR aims to cover, and are these tests written from scratch or taken from an existing implementation (cuDF, Vaex)?

> Is there a plan to make these tests more generic so not only pandas-like dataframes could be tested with them?

That would be quite nice, and was indeed something that multiple maintainers of Pandas and other libraries have asked for.

vnlitvinov commented on protocol/tests/conftest.py:
```diff
@@ -0,0 +1,20 @@
+import pytest
+import pandas as pd
+from pandas.api.exchange.implementation import _from_dataframe
```
this isn't needed

vnlitvinov commented on protocol/tests/conftest.py:
```diff
@@ -0,0 +1,20 @@
+    return maker
+
+
+@pytest.fixture(scope="package")
+def df_from_xchg():
+    def maker(xchg):
+        return _from_dataframe(xchg)
+
+    return maker
```
this is not needed, either - it's a stub for round-trip testing which should be implemented in the library following the spec, not in the spec itself

dchigarev commented on protocol/tests/test_protocol.py:
```diff
@@ -0,0 +1,149 @@
+
+        assert colX.dtype[0] == kind
+
+    assert dfX.get_column_by_name("c").dtype[1] == 64
```
we can't ensure that the `float` column is 64-bit in every dataframe implementation, or can we?

Rubtsowa replied:
@dchigarev I deleted this line

dchigarev commented on protocol/tests/test_protocol.py:
```diff
@@ -0,0 +1,149 @@
+    assert dfX.num_columns() == 3
+    assert dfX.num_rows() == 3
+    assert dfX.num_chunks() == 1
+    assert dfX.column_names() == ["x", "y", "z"]
+    assert (
+        dfX.select_columns((0, 2)).column_names()
+        == dfX.select_columns_by_name(("x", "z")).column_names()
+    )
```
[`.column_names()` returns an iterator](https://data-apis.org/dataframe-protocol/latest/API.html#:~:text=Return%20an%20iterator%20yielding%20the%20column%20names.), so it should probably be converted to a list before comparison
```suggestion
    assert list(dfX.column_names()) == ["x", "y", "z"]
    assert (
        list(dfX.select_columns((0, 2)).column_names())
        == list(dfX.select_columns_by_name(("x", "z")).column_names())
    )
```

Rubtsowa replied:
@dchigarev I changed this

dchigarev commented on protocol/tests/test_protocol.py:
```diff
@@ -0,0 +1,149 @@
+        }
+    )
+    dfX = df.__dataframe__()
+    columns = {"a": 0, "b": 0, "c": 2, "d": 0, "e": 20, "f": 21}
```
TODO: use enums after #74 is merged

vnlitvinov replied:
One cannot do this, as (at least in current agreement) those files (the spec and the tests) are expected to be vendored by the library devs, and we can not restrict them where they should vendor those or how they should name them...

dchigarev commented on protocol/tests/test_protocol.py:
```diff
@@ -0,0 +1,149 @@
+    for colX in dfX.get_columns():
+        assert colX.size == 2
+        assert colX.num_chunks() == 1
+    assert dfX.get_column(0).dtype[0] == 0
+    assert dfX.get_column(1).dtype[0] == 2
```
TODO: use enums after #74 is merged

vnlitvinov replied:
Disagree, see https://github.com/data-apis/dataframe-api/pull/75#discussion_r816729587

YarShev commented:
@Rubtsowa, please leave comments what is fixed.

vnlitvinov commented on protocol/tests/test_protocol.py:
```diff
@@ -36,7 +36,7 @@ def test_float_int(df_from_dict):
     )
     dfX = df.__dataframe__()
-    columns = {"a": 0, "b": 0, "c": 2, "d": 0, "e": 20, "f": 21}
+    columns = [INT, INT, FLOAT, INT, BOOL, STRING]
```
these should be imported from the protocol itself, or at least declared somewhere

Rubtsowa replied:
@vnlitvinov I changed this

Rubtsowa commented:
@rgommers please, check PR

rgommers reviewed (commented):
Thanks for the ping @Rubtsowa. I resolved all the comments that were addressed. This is looking good overall. I do see some failures though. It's a little tricky to run these tests, since this is now a proper pytest test suite with a `conftest.py` but it doesn't apply the needed monkeypatch to attach `__dataframe__` to `pd.DataFrame`. So what I did to run the tests is:
```bash
$ ipython  # in the `protocol/` directory
```
```python
In [1]: run pandas_implementation.py

In [2]: run -m pytest tests/
```
The result is:
```
FAILED tests/test_protocol.py::test_dataframe - pandas.core.indexing.IndexingError: Too ...
FAILED tests/test_protocol.py::test_df_get_chunks[10-3] - assert 1 == 3
FAILED tests/test_protocol.py::test_df_get_chunks[12-3] - assert 1 == 3
FAILED tests/test_protocol.py::test_df_get_chunks[12-5] - assert 1 == 5
FAILED tests/test_protocol.py::test_column_get_chunks[10-3] - assert 1 == 3
FAILED tests/test_protocol.py::test_column_get_chunks[12-3] - assert 1 == 3
FAILED tests/test_protocol.py::test_column_get_chunks[12-5] - assert 1 == 5
FAILED tests/test_protocol.py::test_buffer - TypeError: cannot unpack non-iterable metho...
========================= 8 failed, 8 passed, 1 warning in 0.15s ==========================
```
Do the tests pass for you, and if so are you working from a different version of the protocol implementation for Pandas, or running the tests differently?

Rubtsowa commented:
@rgommers thanks for your comment. 

tests/test_protocol.py::test_dataframe - In this test I see problem with function ``select_columns``. I have error: 
```
def select_columns(self, indices: Sequence[int]) -> '_PandasDataFrame':
>       if not isinstance(indices, collections.Sequence):
E       AttributeError: module 'collections' has no attribute 'Sequence'
```

tests/test_protocol.py::test_df_get_chunks, tests/test_protocol.py::test_column_get_chunks - Pandas do not plan to use more chunks than 1. Because ``assert len(chunks) == n_chunks`` will be don't correct. I fixed this.

tests/test_protocol.py::test_buffer - In this test was don't correct assignment. I fixed this.

Rubtsowa commented:
@rgommers thanks for your comment. 

tests/test_protocol.py::test_dataframe - In this test I see problem with function ``select_columns``. I have error: 
```
def select_columns(self, indices: Sequence[int]) -> '_PandasDataFrame':
>       if not isinstance(indices, collections.Sequence):
E       AttributeError: module 'collections' has no attribute 'Sequence'
```

tests/test_protocol.py::test_df_get_chunks, tests/test_protocol.py::test_column_get_chunks - Pandas do not plan to use more chunks than 1. Because ``assert len(chunks) == n_chunks`` will be don't correct. I fixed this.

tests/test_protocol.py::test_buffer - In this test was don't correct assignment. I fixed this.

dchigarev commented on protocol/tests/test_protocol.py:
```diff
@@ -0,0 +1,163 @@
+def test_df_get_chunks(size, n_chunks, df_from_dict):
+    df = df_from_dict({"x": list(range(size))})
+    dfX = df.__dataframe__()
+    chunks = list(dfX.get_chunks(n_chunks))
+    """
+    Pandas do not plan to use more chunks than 1. 
+    But if chunking is supported in your implementation, 
+    then it is the place to use the following check
+    
+    ``assert len(chunks) == n_chunks``
+    
+    """
+    assert len(chunks) == 1
```
But we explicitly specify the number of chunks that we desire to recieve for the `.get_chunks()`  via `n_chunks` parameter. Going by the protocol's spec, if the parameter is specified, _the producer must subdivide each chunk before yielding it_ [[1]](https://github.com/data-apis/dataframe-api/blob/81be345bef2c9522ae178bf4bd7bf788a50056d6/protocol/dataframe_protocol.py#L382-L390). So it doesn't mater whether an underlying implementation internally chunks data or not, it still has to return `n_chunks` amount of chunks when we explicitly request it.

The reason why @rgommers experienced failures [[2]](https://github.com/data-apis/dataframe-api/pull/75#pullrequestreview-911993634) for this test, is that the draft protocol's implementation in `pandas_implementation.py` doesn't respect `n_chunks` parameter:
https://github.com/data-apis/dataframe-api/blob/81be345bef2c9522ae178bf4bd7bf788a50056d6/protocol/pandas_implementation.py#L762-L766

The latest known pandas implementation for the protocol (pandas-dev/pandas#46141) do respect this parameter and _should_ work properly with the previous version of this test (I haven't tested it for myself)

YarShev replied:
I wonder if a library doesn't support physical chunking (i.e. the column of a dataframe is a continuous block of memory), should that library implement virtual chunking to support get_chunks()?

cc @rgommers, @vnlitvinov

rgommers replied:
Thanks for the explanations @Rubtsowa and @dchigarev. That makes perfect sense.

> should that library implement virtual chunking to support get_chunks()?

Yes indeed.

rgommers replied:
This explanation is right, so the last commit should be reverted I think.

YarShev commented:
> Do the tests pass for you, and if so are you working from a different version of the protocol implementation for Pandas, or running the tests differently?

I would like the tests from this PR fit every implementation of the protocol, but not the implementation for pandas only.

rgommers commented:
> I would like the tests from this PR fit every implementation of the protocol, but not the implementation for pandas only.

They should indeed. That was not the point of the comment though, I just had the wrong (in-progress) implementation of the protocol for Pandas. We should just link to https://github.com/pandas-dev/pandas/pull/46141, I think that's the right implementation and not the one in this repo.

Rubtsowa commented:
Then Failed in tests which have @rgommers at the moment they will be present, but with the updated ramp, they will not occur.

honno commented:
So [`dataframe-interchange-tests`](https://github.com/data-apis/dataframe-interchange-tests) now covers everything in this PR, _except_ for asserting the actual data held in buffers e.g.

```python
for idx, truth in enumerate(arr):
    val = ctype.from_address(dataBuf.ptr + idx * (bitwidth // 8)).value
    assert val == truth, f"Buffer at index {idx} mismatch"
```

I can mull doing something like this, similarly only in scenarios where checking the actually values of a buffer is easy (e.g. on CPU, and only when libraries are returning valid `Column.dtype` tuples in the first place).

rgommers commented:
Thanks @honno. I'll close this then. Can you migrate that last loose end to a new issue (or PR) on the `dataframe-interchange-tests` repo so it doesn't get lost?

Thanks a lot @Rubtsowa and @dchigarev. Given that:
- after gh-81 the implementation (including all the existing tests) has been removed from this repo,
- Pandas now has its own implementation merged with significant improvements compared to the prototype, and
- https://github.com/data-apis/dataframe-interchange-tests covers protocol testing comprehensively now,
I will close this PR.

Thanks for opening this PR! And also for working on the Pandas prototype (I think).
